host/traceh.c

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/trace.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#if defined(__linux__)
+#include <sys/time.h>
+#endif
+#include <time.h>
+#include "enclave.h"
+#include "hostthread.h"
+
+#define LOGGING_FORMAT_STRING "%02d:%02d:%02d:%06ld TID(0x%lx) (%s)[%-5s]%s"
+static char* log_level_strings[OE_LOG_LEVEL_MAX] = {"",
+                                                    "FATAL",
+                                                    "ERROR",
+                                                    "WARN",
+                                                    "INFO"};
+static oe_mutex g_log_lock = OE_H_MUTEX_INITIALIZER;
+static const char* g_log_file_name = NULL;
+static log_level_t g_log_level = OE_LOG_LEVEL_ERROR;
+static bool g_initialized = false;
+
+static log_level_t env2log_level(void)
+{
+    log_level_t level = OE_LOG_LEVEL_ERROR;
+    const char* level_str = getenv("OE_LOG_LEVEL");
+
+    if (level_str == NULL)
+    {
+        goto done;
+    }
+    else if (strcmp(level_str, "INFO") == 0)
+    {
+        level = OE_LOG_LEVEL_INFO;
+    }
+    else if (strcmp(level_str, "WARNING") == 0)
+    {
+        level = OE_LOG_LEVEL_WARNING;
+    }
+    else if (strcmp(level_str, "ERROR") == 0)
+    {
+        level = OE_LOG_LEVEL_ERROR;
+    }
+    else if (strcmp(level_str, "FATAL") == 0)
+    {
+        level = OE_LOG_LEVEL_FATAL;
+    }
+    else if (strcmp(level_str, "NONE") == 0)
```

Comment:
Nonblocking: Was this just added? It's unusual because this was previously handled by the `level_str == NULL` case and we don't have a string represenation of this in the `log_level_strings` (They don't have to match since it'll never be used, but it's odd that we would recognize this specifically as a keyword.) Also, this reminded me: is empty string a case you need to handle (variable defined, not assigned a value)